website/src/pages/tech-specs.md

```diff
@@ -0,0 +1,371 @@
+3. For the merge on read table type, ensure the appropriate merging rules are applied to apply the updates queued for the base in the log files.
+   1. Contents of the log files should be loaded into an effective point lookup data structure (in-memory or persisted)
+   2. Duplicate row keys should be merged based on the ordering field specified. It is important to order the inserts and deletes in the right order to be consistent and idempotent. 
+   3. When the base file is scanned, for every row block, the reader has to lookup if there is a newer version of the data available for the row keys in the block and merge them into the record iterator. 
+
+
+
+## Writer Expectations
+
+Writer into Hudi will have to ingest new records, updates to existing records or delete records into the dataset. All transactional actions follow the same state transition as described in the transaction log (timeline) section. Writers will optimistically create new base ang log files and will finally transition the action state to completed to register all the modifications to the dataset atomically. Writer merges the data using the following steps
+
+1. Writer will pick a monotonically increasing instant time from the latest state of the Hudi timeline (**action commit time**) and will pick the last successful commit instant (**merge commit time**) to merge the changes to. If the merge succeeds, then action commit time will be the next successful commit in the timeline. 
+2. For all the incoming records, the writer will have to efficiently determine if this is an update or insert. This is done by a process called tagging - which is a batched point lookups of the row key and partition path pairs in the entire dataset. The efficiency of tagging is critical to the merge performance. This can be optimized with indexes (bloom, global key value based index) and caching. New records will not have a tag. 
+3. Once records are tagged, the writer can apply them onto the specific file slice. 
+   1. For copy on write, writer will create a new slice (action commit time) of the base file in the file group
+   2. For merge on read, writer will create a new log file with the action commit time on the merge commit time file slice
+4. Deletes are encoded as special form of updates where only the meta fields and the operation is populated. See the delete block type in log format block types. 
+5. Once all the writes into the file system is complete, concurrency control checks happen to ensure there are no overlapping writes and if that succeeds, the commit action is completed in the timeline atomically making the changes merged visible for the next reader.  
+6. Synchronizing Indexes and metadata needs to be done in the same transaction that commits the modifications to the table. 
+
+
+
+## Table maintenance
```

Comment:
should we call this table services?

Reply:
should also include archive, rollback, savepoint ?

Reply:
Here the call out is how to maintenance over this storage spec. Table services is how we have done in the Hudi platform. Hudi platform is an implementation of the spec.